### PR TITLE
Dispatch fixes for #421

### PR DESCRIPTION
**Pull request by ReeceHumphreys**
Includes source changes, will add tests this weekend.

### CODE REVIEW

File: swift/src/Ice/Incoming.swift [hunk 1]
```diff
@@ -92,6 +98,22 @@ public final class Incoming {
         self.format = format
     }
 
+    public func setResultPromise<T>(_ p: Promise<T>,
+                                    _ cb: ((OutputStream, T) -> Void)? = nil) -> Promise<OutputStream> {
```
Comment (bernardnormier) [inline swift/src/Ice/Incoming.swift]: I prefer the way Ice for Java does this, with two overloaded methods, one for Promise<T> with T != Void and a non-optional closure, and one for Promise<Void> and no closure at all.

File: swift/src/Ice/Incoming.swift [hunk 2]
```diff
@@ -98,18 +91,43 @@ public final class Incoming {
...
+    }
+
+    public func setResult() -> Promise<OutputStream>? {
+        let ostr = OutputStream(communicator: istr.communicator, encoding: current.encoding)
+        ostr.writeEmptyEncapsulation(current.encoding)
+        self.ostr = ostr
+        return nil // Response is cached in the Incoming to not have to create unnecessary future
+    }
+
+    public func setResult(_ cb: @escaping (OutputStream) -> Void) -> Promise<OutputStream>? {
```
Comment (pepone) [inline swift/src/Ice/Incoming.swift]: do we really need `@escaping` for this callback?

File: swift/src/Ice/Incoming.swift [hunk 3]
```diff
@@ -98,18 +91,43 @@ public final class Incoming {
         self.format = format
     }
 
+    @discardableResult
+    public func setResult(_ os: OutputStream) -> Promise<OutputStream>? {
+        ostr = os
+        return nil // Response is cached in the Incoming to not have to create unnecessary future
```
Comment (pepone) [inline swift/src/Ice/Incoming.swift]: This comments about future, should be about promise

File: swift/src/Ice/Incoming.swift [hunk 4]
```diff
@@ -170,10 +188,13 @@ public final class Incoming {
             if let promise = try s.dispatch(request: self, current: current) {
                 // dispatched asyncroniasdfly
```
Comment (pepone) [inline swift/src/Ice/Incoming.swift]: There is a typo there

File: swift/src/Ice/BlobjectAsync.swift [hunk 5]
```diff
@@ -33,14 +33,10 @@ public struct BlobjectAsyncDisp: Disp {
         self.servant = servant
     }
 
-    public func dispatch(request: Request, current: Current) throws {
+    public func dispatch(request: Request, current: Current) throws -> Promise<OutputStream>? {
```
Comment (pepone) [inline swift/src/Ice/BlobjectAsync.swift]: Need to update the doc comment to mention the return type

**Pull request merged**